src/persist/src/indexed/mod.rs

```diff
@@ -1308,6 +1469,254 @@ mod tests {
+        let error_response =
+            MaintenanceRes::CompactTrace((id, Err(Error::from("test compaction error"))));
+        i.handle_maintenance_res(error_response.clone());
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving an error, retry the original request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request);
+
+        // Handle receiving a valid MaintenanceRes.
+        let response =
+            futures_executor::block_on(request.clone().to_future(&maintainer).run_async());
+
+        i.handle_maintenance_res(response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving a valid response, issue a new compaction request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request2 = reqs[0].clone();
+        assert!(request2 != request);
+
+        // Construct a new valid, response.
+        let response2 =
+            futures_executor::block_on(request2.clone().to_future(&maintainer).run_async());
+
+        // Send back an error response. After this there are no requests in flight.
+        i.handle_maintenance_res(error_response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // Retry request after error response.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request2);
+
+        // Handle failed write to metadata after receiving a response.
+        i.handle_maintenance_res(response2.clone());
+        unreliable.make_unavailable();
+        assert_eq!(i.drain_pending(), Err(Error::from("unavailable: blob set")));
+        unreliable.make_available();
+
+        // Retry request after failed meta write.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request2);
+
+        // Handle receiving a response after the trace is deleted.
+        block_on(|res| i.destroy("0", res))?;
+
+        // Handle receiving an valid MaintenanceRes for a stream that has been deleted.
+        i.handle_maintenance_res(response2);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // Don't produce any more requests because the stream is deleted.
+        let reqs = i.step()?;
+        assert_eq!(reqs, vec![]);
+
+        Ok(())
+    }
+
+    /// This test checks that Indexed correctly handles receiving an invalid MainenanceRes
+    /// (one whose request does not match the currently in-flight request). This test is
+    /// expected to panic when run with cfg!(debug_assertions).
+    fn trace_compaction_invalid_response() -> Result<(), Error> {
+        let unreliable = UnreliableHandle::default();
+        let (mut i, maintainer, _) = trace_compaction_setup(unreliable)?;
+
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request = reqs[0].clone();
+
+        // Handle receiving an invalid (corrupt) MaintenanceRes.
```

Comment:
yeah corrupt also means something specific to me that doesn't really apply here :). I think you and I have different takes on how "fundamental" the infallibility of the connection between indexed and maintainer are. I've come to agree with your argument that we shouldn't handle unexpected responses in indexed yet, but I don't want to bake this in too much